### PR TITLE
Add initial github module

### PR DESCRIPTION
**Pull request by adrianmoisey**
##### Issue Type:
- New Module Pull Request
##### Plugin Name:

github
##### Ansible Version:

```
ansible 2.1.0 (devel f99ed97c40) last updated 2016/02/28 20:44:14 (GMT +200)
  lib/ansible/modules/core:  not found - use git submodule update --init lib/ansible/modules/core
  lib/ansible/modules/extras:  not found - use git submodule update --init lib/ansible/modules/extras
  config file =
  configured module search path = Default w/o overrides
```
##### Summary:

This adds a simple module for fetching the latest release of a repo or list the available branches from Github. This could be useful for someone wanting to do releases from Github.
This module could be extended a bit more. For example: it could possibly create pull requests.

### CODE REVIEW

**Comment by sivel** [comment]
I would rather see this named `github_release` and only handle release related tasks.  Modules should generally be targeted in functionality.

**Comment by adrianmoisey** [comment]
That's a good suggestion. I wasn't sure what direction the module should take when I started it. I'll tidy it up a bit and push that change.

**Comment by adrianmoisey** [comment]
I've refactored it and made it release specific. Currently it only gets the latest release.

**Comment by sivel** [comment]
```
============================================================================
./source_control/github_release.py
============================================================================
ERROR: DOCUMENTATION.options.action.description: required key not provided
```

**Comment by adrianmoisey** [comment]
I've updated the PR and fixed the error. Thanks @sivel

File: source_control/github_release.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible. If not, see <http://www.gnu.org/licenses/>.
+
+
+DOCUMENTATION = '''
+---
+module: github
```
Comment (gundalow) [inline source_control/github_release.py]: I believe this needs changing to github_release

File: source_control/github_release.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        description:
+            - Repository name
+    action:
+        required: true
+        description:
+            - Action to perform
+        choices: [ 'latest_release' ]
+
+author:
+    - "Adrian Moisey"
```
Comment (gundalow) [inline source_control/github_release.py]: You may wish to put your github here as well

`"Adrian Moisey (@adrianmoisey)"`

**Comment by gundalow** [comment]
pep8, pyflakes and ansible-validate-modules are all clean
- All functions are documented
- Correctly deals with error cases

The release function of GitHub isn't something I've used, so I don't know how useful this would be for most people. Though code and implementation looks sensible

**Comment by adrianmoisey** [comment]
Thank for the review. I've updated as per your comments.

**Comment by gundalow** [comment]
Thank you for the quick turn around for those fixes.

Confirmed that the issues that I'd raised have been fixed. 

I've not done any functional testing of this yet...

**Comment by gundalow** [comment]
Hi, if you think all the outstanding comments have been addressed, please comment with text `ready_for_review` and we will put this PR back into review.

**Comment by adrianmoisey** [comment]
> Hi, if you think all the outstanding comments have been addressed, please comment with text ready_for_review and we will put this PR back into review.

Thanks. I think all comments have been addressed, unfortunately i don't have permissions to change any labels on this PR.

**Comment by adrianmoisey** [comment]
I didn't read your comment correctly, sorry.

**Comment by adrianmoisey** [comment]
ready_for_review

File: source_control/github_release.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    if action == 'latest_release':
+        release = repository.latest_release()
+        if release:
+            module.exit_json(tag=release.tag_name)
+        else:
+            module.exit_json(tag=None)
+
+
+from ansible.module_utils.basic import *
+main()
```
Comment (gundalow) [inline source_control/github_release.py]: Hi, 

This should be

if **name** == '**main**':
  main()

Comment (adrianmoisey) [inline source_control/github_release.py]: Thanks, I've fixed that in https://github.com/ansible/ansible-modules-extras/pull/1755/commits/ee78d08566a16bc48379e2cda44fac6ef3e027b1

**Comment by gregdek** [comment]
Thanks @adrianmoisey for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by adrianmoisey** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @adrianmoisey for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by adrianmoisey** [comment]
:bell:

**Comment by gundalow** [comment]
Note to self: gundalow_review_this

**Comment by gundalow** [comment]
Hi,

Looks like there is a small tweak needed to support Python 3

```
python3 -m compileall -fq github_release.py 
*** Error compiling 'github_release.py'...
  File "github_release.py", line 101
    except github3.AuthenticationFailed, e:
                                       ^
SyntaxError: invalid syntax

```

The following guide should explain what's needed http://docs.ansible.com/ansible/developing_modules_python3.html

Thanks

**Comment by gundalow** [comment]
needs_revision

Just one thing to fix then this can be merged.

**Comment by adrianmoisey** [comment]
All fixed, thanks. I rebased against `devel`, as Travis was complaining about some conflicts.

**Comment by gundalow** [comment]
@adrianmoisey Thank you for the quick turn around.

shipit

**Comment by gregdek** [comment]
Thanks again to @adrianmoisey for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request merged**